src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -102,19 +102,21 @@ public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
     	CacheService cacheService = new CacheService(dto);
     	MessageCacheItem cacheItem = cacheService.getCacheOfComponent();
     	if (cacheItem != null) { // Item is in cache
-    		if (cacheItem.getCachedData().isEmpty()) { // This means that the data to be used is from a fallback locale.
-				// If expired, try to first create and store cacheItem for the requested locale in a separate thread.
-				if (cacheItem.isExpired())
-					this.createCacheItemTask(null); // Pass null so that locale fallback will not be applied.
-
-    			// Use cached data of cacheItem.getLocale() --> fallback locale
-				MessagesDTO fallbackLocaleDTO = new MessagesDTO(dto.getComponent(), cacheItem.getLocale(), dto.getProductID(), dto.getVersion());
-				cacheItem = new ComponentService(fallbackLocaleDTO).getMessages(null);
-			} else if (cacheItem.isExpired()) {
+    		if (cacheItem.isExpired()) {
     			// Refresh the cacheItem in a separate thread
     			refreshCacheItemTask(cacheItem);
     		}
-    	} else { // Item is not in cache. Create and store cacheItem for the requested locale
+    	} else { // Item is not in cache.
+			ProductService ps = new ProductService(dto);
+			if (!ps.isSupportedLocale(Locale.forLanguageTag(dto.getLocale()))) { // Requested locale is not supported
+				Locale matchedLocale = LocaleUtility.pickupLocaleFromList(new LinkedList<>(ps.getSupportedLocales()), Locale.forLanguageTag(dto.getLocale()));
+				if (ps.isSupportedLocale(matchedLocale)) { // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
```

Comment:
Can't imagine getSupportedLocales is invoked three times in three code lines.

Reply:
I think you can improve the LocaleUtility.pickupLocaleFromList to get enough information by only one call. Picking up once to get the supported locale or empty locale.

Reply:
Ahhhm yes! I agree. I already have that lined up. However, as I said, this PR is only to clean up / remove any code related to adding an empty data set into the cache.